### PR TITLE
List patterns: parsing changes

### PR DESCRIPTION
**Pull request by alrz**
In this PR:

- Add langversion check in the parser
- Relax slice subpattern to accept any pattern
- Support length patterns in combinators

Test plan: https://github.com/dotnet/roslyn/issues/51289

### CODE REVIEW

**Comment by alrz** [comment]
Turns out this affects other tests we just changed in the other PR. So this should be reviewed afterwards.

File: src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests_ListPatterns.cs [hunk 1]
```diff
@@ -673,6 +673,146 @@ public void LengthPattern_16()
             EOF();
         }
 
+        [Fact]
+        public void LengthPattern_17()
+        {
+            UsingExpression(@"c is not [<0]");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests_ListPatterns.cs:679]: Let's add full tests for those scenarios that now pass (verifying diagnostics and execution).

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 3)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7)

**Comment by jcouv** [comment]
@333fred @dotnet/roslyn-compiler for second review. Thanks

**Comment by alrz** [comment]
Closing in favor of #54335

**Pull request closed without merging**